GraphQL-Schema-Design.md

```diff
@@ -0,0 +1,142 @@
+        }
+      }
+    }
+  }
+  ```
+
+## Root fields
+
+* These are entry points into the graph that is Artsy’s data set.
+
+* Not all types necessarily make sense as root fields.
+  - In general only domain models should be exposed as root fields. An ‘artwork’ is such a model, Artsy’s system time
+    (for clock synchronization) probably is not.
+
+* For the cases where there is no root field, but you still need to be able to retrieve an arbitrary node of the graph
+  (e.g. when you need to refetch a node without needing to refetch all parent nodes along the path from the root to said
+  node), there is the special [`node` root field](https://facebook.github.io/relay/graphql/objectidentification.htm).
+
+  In short, this field is able to retrieve an arbitrary node by using a special ‘global’ ID, which has data encoded
+  needed for our system to know the type of that entity and how to retrieve it. For example, based on just
+  `banksy-champagne-formica-flag` the system wouldn’t be able to know what type of entity this refers to, a global ID
+  would rather encode it like `Artwork:banksy-champagne-formica-flag`.
+
+## Lists
+
+It is undesirable to have multiple fields that semantically refer to the same data. So rather than defining both e.g. an
+`artworks` field _and_ an `artworks_connection`, simply choose one form applicable to the data and call it `artworks`.
+
+* A paginated list (such as most associated types) should use ‘connections’ (see
+  [the spec](https://facebook.github.io/relay/graphql/connections.htm) and
+  [this blog post](https://dev-blog.apollodata.com/explaining-graphql-connections-c48b7c3d6976)).
+  - These are more forward-compatible, as they allow for adding metadata to the association itself and the ‘edge’ (the
+    relation between the parent entity and the associated entity).
+  - Connections use
+    [cursor based pagination](https://www.sitepoint.com/paginating-real-time-data-cursor-based-pagination/), which has
+    benefits over ‘page’ based pagination in systems where data availability could change (e.g. removing/adding of
+    artworks) and it’s an abstraction that allows the underlying system to change the way it does pagination whilst
+    staying API compatible.
+
+* In rare cases an immediate list may be used, but this should only be done in cases where the list has few entries and
+  doesn’t require pagination, i.e. fetch _all_ entries at once.
+
+## Schema stitching
+
+As we expand our microservices oriented architecture and take our use of GraphQL more serious, a need has arisen to
+model GraphQL schemas more closely to the data source (i.e. database) in an effort to colocate schema code next to the
+rest of the code related to that data and to make these schema resolvers more performant.
+
+In this new scenario, metaphysics would be an orchestration layer that [stitches together](https://www.apollographql.com/docs/graphql-tools/schema-stitching.html) these various schemas into a single coherent
```

Comment:
Everything is in #graphql and they are really nothing but the library code still being young and we need to expand on it for our case. I wouldn’t call these difficulties, they are hurdles that are to be expected.

Reply:
Regarding your second point, totally! The name ‘stitching’ is just what they have been calling it, but in reality there’s really more composition that we’ll be doing rather than just merging schemas to live side by side.